src/runtime/story.rs

```diff
@@ -0,0 +1,192 @@
+    where
+        Output: Write,
+    {
+        let object = self.peek_cursor().ok_or(())?;
+        if self.execute(object.clone()) {
+            self.advance_cursor()?;
+        }
+        Ok(())
+    }
+
+    fn execute(&mut self, object: RuntimeObject) -> bool
+    where
+        Output: Write,
+    {
+        match object {
+            RuntimeObject::Choice(_choice) => todo!(),
+            RuntimeObject::Container(container) => {
+                // push first index of container to cursor stack
+                self.cursors.push((container.clone(), 0));
+                // shouldn't advance cursor
+                return false;
+            }
+            RuntimeObject::ControlCommand(command) => self.command(command),
+            RuntimeObject::Divert(_divert) => todo!(),
+            RuntimeObject::Glue => self.glue = true,
+            RuntimeObject::NativeFunctionCall(_call) => todo!(),
+            RuntimeObject::Tag(_tag) => todo!(),
+            RuntimeObject::Value(value) => match value {
+                Value::Int(_) => todo!(),
+                Value::Float(_) => todo!(),
+                Value::String(string) => self.output(string),
+                Value::DivertTarget { target_path: _ } => todo!(),
+                Value::VariablePointer {
+                    name: _,
+                    context_index: _,
+                } => todo!(),
+            },
+            RuntimeObject::VariableAssignment(_assignment) => todo!(),
+            RuntimeObject::VariableReference(VariableReference { name: _ }) => todo!(),
+            RuntimeObject::ReadCount(ReadCount { target: _ }) => todo!(),
+            RuntimeObject::Void => {}
+        }
+        // should advance cursor
+        true
+    }
+
+    fn output<Object>(&mut self, object: Object)
+    where
+        Object: fmt::Display,
+        Output: Write,
+    {
+        if self.glue {
+            todo!()
+        }
+        self.output_text
+            .write_str(&format!("{}", object))
+            .expect("Error writing Ink output to output stream");
+    }
+
+    fn command(&mut self, command: ControlCommand) {
+        match command {
+            ControlCommand::Done => {
+                // TODO close current thread
+            }
+            _ => todo!(),
+        }
+    }
+
+    fn peek_cursor(&self) -> Option<&RuntimeObject> {
+        let (container, index) = self.cursors.last()?;
+        let object = container.content.get(*index)?;
+        Some(object)
+    }
+
+    fn advance_cursor(&mut self) -> Result<(), ()> {
+        let (mut container, mut index) = self.cursors.pop().ok_or(())?;
+        while container.content.len() <= index + 1 {
+            // pop all finished containers from stack
+            (container, index) = self.cursors.pop().ok_or(())?
+        }
+        // place index back on cursor stack, incremented
+        self.cursors.push((container, index + 1));
+        Ok(())
+    }
+}
+
+impl<Output> Story<Output> {
```

Comment:
Is there a reason this has a separate `impl` block?

Reply:
Oops, that's probably a holdover from the previous commit where they had different generics.